### ISSUE TITLE
Add import metadata to JSON history

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#460]
_Original author: mccus...@gmail.com (October 12, 2011 03:59:50)_

## What steps will reproduce the problem?
1. Load some data.
1. Do some manipulations
1. Re-apply those manipulations to a new dataset or new version of a dataset.

Currently, it's not possible to re-apply a set of manipulations without exporting, messing with the export file, and re-importing. It would be extremely useful to provide repeatable transformation capabilities. Without these, the use of Refine for scientific (repeatable) applications is extremely limited.

_Original issue: http://code.google.com/p/google-refine/issues/detail?id=460_

**Comment by stevenqzhang** [comment]
Any update on this issue? It's been 4 years...

**Comment by thadguidry** [comment]
@stevenqzhang The VIB-Bits plugin is what most folks use to solve this issue.  In fact, we could probably close this issue, since the general use case is handled nicely by the plugin.

**Comment by PointyShinyBurning** [comment]
The VIB-Bits plugin doesn't quite cover it for me, as I'd like to make repeatable all the steps including the decisions made in initial data import about formats, columns to skip etc.

It would make most sense for the initial data load to appear like a normal operation in the edit history to be replayed, exported as JSON etc. Is there any interest in work towards this approach?

**Comment by wetneb** [comment]
> It would make most sense for the initial data load to appear like a normal operation in the edit history to be replayed, exported as JSON etc. Is there any interest in work towards this approach?

Yes it has been proposed before to add import metadata to the history. I think it would be an obvious move.

**Comment by wetneb** [comment]
Here is an in-the-wild example of an OpenRefine workflow being shared, where the import settings are described externally:
https://github.com/OCLC-Developer-Network/WikidataHoldingsMatching

**Comment by VladimirAlexiev** [comment]
https://github.com/OpenRefine/OpenRefine/issues/5164 was posted by @nikolatulechki of Ontotext. In the context of OntoRefine, we have discussed several **reproducibility scenarios**:
1. **Run project with settings**: take import options JSON, input file(s), cleaning ops (operation history JSON), SPARQL transformation (applies to OntoRefine only). Create new project, apply cleaning, apply transformation, apply SPARQL transform, save output, optionally delete this new project.
This is Use Case “Data Engineer 1 makes the transform and mapping and gives it to Data Engineer 2 to automate and execute it”.
It's similar to what eg `tarql` does
2. **Run project from template project**: take a project_ID and input file(s), create new project, apply the original project's import options, apply cleaning, apply transformation, optionally delete this new project. 
This is use case “Data Engineer 1 makes the transform and mapping using a small sample file, then wants to run it on a big file, but keep the small project for further editing and refining of the cleaning/transform”.
3. **Rerun project with new data**: take a project_ID and input file(s), create new project, apply the original project's import options, apply cleaning, apply transformation, delete the old project. 
This is use case “Data Engineer 1 makes the transform and mapping on the first portion of data, then wants to run it on a second portion, so doesn’t need to keep the old project”


- It is already possible for CLI writers to implement the above scenarios:
OpenRefine import options can be obtained by API `get-project-metadata`, field `importOptionMetadata` (as can be seen in “Open Project>About”).
- Storing the import options (or why not all metadata?) in the operation history will simplify scenario 1 because there'll be one JSON not two
- But [create project](https://docs.openrefine.org/technical-reference/openrefine-api#create-project) and [apply operations](https://docs.openrefine.org/technical-reference/openrefine-api#apply-operations) are two separate APIs. I don' think you can change the import options after the project is created. So we'd need a new API "create project and apply operations"
